src/main/java/co/cask/spanner/source/SpannerSource.java

```diff
@@ -0,0 +1,259 @@
+        case FLOAT64:
+          builder.set(fieldName, resultSet.getDouble(fieldName));
+          break;
+        case STRING:
+          builder.set(fieldName, resultSet.getString(fieldName));
+          break;
+        case BYTES:
+          ByteArray byteArray = resultSet.getBytes(fieldName);
+          builder.set(fieldName, byteArray.toByteArray());
+          break;
+      }
+    }
+    emitter.emit(builder.build());
+  }
+
+  @Override
+  public void onRunFinish(boolean succeeded, BatchSourceContext context) {
+    LOG.info("Run finished, closing spanner client");
+    // free up spanner resources
+    if (spanner != null) {
+      spanner.close();
+    }
+  }
+
+  private void initializeConfig(Configuration configuration) {
+    setIfValueNotNull(configuration, SpannerConstants.PROJECT_ID, config.project);
+    setIfValueNotNull(configuration, SpannerConstants.SERVICE_ACCOUNT_FILE_PATH, config.serviceAccountFilePath);
+    setIfValueNotNull(configuration, SpannerConstants.INSTANCE_ID, config.instance);
+    setIfValueNotNull(configuration, SpannerConstants.DATABASE, config.database);
+    setIfValueNotNull(configuration, SpannerConstants.QUERY, config.query);
+  }
+
+  private void setIfValueNotNull(Configuration configuration, String key, String value) {
+    if (value != null) {
+      configuration.set(key, value);
+    }
+  }
+
+  /**
+   * Serialize the object into bytes and encode the bytes into string using Base64 encoder.
+   * @throws IOException
+   */
+  private String getSerializedObjectString(Object object) throws IOException {
+    try (ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
+         ObjectOutputStream objectOutputStream = new ObjectOutputStream(byteArrayOutputStream)) {
+      objectOutputStream.writeObject(object);
+      objectOutputStream.flush();
+      return Base64.getEncoder().encodeToString(byteArrayOutputStream.toByteArray());
+    }
+  }
+
+  private PartitionOptions getPartitionOptions() {
+    PartitionOptions.Builder builder = PartitionOptions.newBuilder();
+    if (config.partitionSizeMB != null) {
+      builder.setPartitionSizeBytes(config.partitionSizeMB * 1024 * 1024);
+    }
+    if (config.maxPartitions != null) {
+      builder.setMaxPartitions(config.maxPartitions);
+    }
+    return builder.build();
+  }
+
+  /**
+   * Endpoint method to get the output schema of spanner query.
+   *
+   * @param request {@link SpannerSourceConfig} containing information required for query to execute.
+   * @return schema of fields
+   */
+  @Path("getSchema")
+  public Schema getSchema(SpannerSourceConfig request) throws Exception {
+    Spanner spanner = SpannerUtil.getSpannerService(request.serviceAccountFilePath, request.project);
+    DatabaseClient databaseClient =
+      spanner.getDatabaseClient(DatabaseId.of(request.project, request.instance, request.database));
+    ResultSet resultSet = databaseClient.singleUse().executeQuery(Statement.of(request.query));
```

Comment:
is the user charged for the full query here even though we don't go through the whole result set?

Reply:
also, do you need to close anything like ResultSet, DatabaseClient, or Spanner?

Reply:
my bad. fixed to close the resources now. about the pricing, https://cloud.google.com/spanner/pricing - documentation says its based on network bandwidth utilization for queries. so i reckon it will be based on result set utilization, but i can confirm by performing some checks.

Reply:
I see. Also, did you figure out how to check if the query can be partitioned?

Reply:
can you also make sure errors are handled here? I remember there's a specific type of exception that is supposed to be throw, otherwise the message won't make it to the response (and to the UI)

Reply:
If the plugin endpoint throws an exception, `Method#invoke` will throw an `InvocationTargetException` and in ArtifactHttpHandler we handle `InvocationTargetException`. So i don't think that issue exists currently. I was able to test with an exception case.

Reply:
however i will update to throw more meaningful exception if the result set is empty.